File: samples/typescript/01-single-call.ts
// ==============================================================================
// 01 — Single Call (TypeScript/JS-Client)
// ==============================================================================
// Ein einzelner RPC-Aufruf. Gezeigt für REST und WebSocket — wähle je nach
// Anforderung. WebSocket ist der empfohlene primäre Kanal (persistent, geringe
// Latenz); REST ist zustandslos und am einfachsten.
//
// Hinweis: Der TS-WebSocket-Client nimmt die Basis-URL (https://…) + einen
// separaten wsPath (Default "sleipnirws"); er hebt intern auf wss://…/sleipnirws ab.
// ==============================================================================

import { SleipnirRestClient, SleipnirCall, SleipnirWebSocketClient } from "sleipnir-client";

type Customer = { id: number; name: string; email: string };

export async function run(rest: SleipnirRestClient): Promise<void> {
  // --- REST: Kunden anlegen (skalarer int kommt zurück) ------------------------
  const newId = await rest.callJson<number>("Customer", "AddCustomer", {
    name: "Alice",
    email: "alice@x.com",
  });
  console.log(`  [REST]    AddCustomer -> neue Id = ${newId}`);

  // --- REST: Kunden laden (callJson<T> deserialisiert data direkt nach T) ------
  const c = await rest.callJson<Customer>("Customer", "GetCustomerById", { id: newId });
  console.log(`  [REST]    GetCustomerById(${newId}) -> ${c?.name} <${c?.email}>`);

  // --- REST: alle Kunden (Liste) ----------------------------------------------
  const all = await rest.callJson<Customer[]>("Customer", "GetAllCustomers");
  console.log(`  [REST]    GetAllCustomers -> ${all?.length ?? 0} Kunde(n)`);

  // --- WebSocket: derselbe Aufruf über den persistenten Kanal -----------------
  // baseUrl + Default-wsPath "sleipnirws" -> wss://localhost:5001/sleipnirws.
  const ws = new SleipnirWebSocketClient("https://localhost:5001");
  await ws.connect();
  try {
    const c2 = await ws.callJson<Customer>(
      SleipnirCall.init("Customer", "GetCustomerById").with({ id: newId }).toRequest(),
    );
    console.log(`  [WebSocket] GetCustomerById(${newId}) -> ${c2?.name} <${c2?.email}>`);
  } finally {
    ws.close();
  }

  // --- Raw-Form (ohne Fluent Builder) — gelegentlich nützlich -----------------
  // params = SleipnirParameter[]; data ist der native JSON-Wert. GetAllCustomers hat
  // keine Parameter -> leeres Array.
  const rawList = await rest.callJson<Customer[]>({
    controller: "Customer",
    method: "GetAllCustomers",
    id: "Customer.GetAllCustomers",
    params: [],
  });
  console.log(`  [REST raw] GetAllCustomers -> ${rawList?.length ?? 0} Kunde(n)`);
}

File: samples/csharp/01-single-call.cs
// ==============================================================================
// 01 — Single Call (C#-Client)
// ==============================================================================
// Ein einzelner RPC-Aufruf. Gezeigt für REST und WebSocket — wähle je nach
// Anforderung. WebSocket ist der empfohlene primäre Kanal (persistent, geringe
// Latenz); REST ist zustandslos und am einfachsten.
// ==============================================================================

using System.Text.Json;
using SleipnirClient.Sleipnir;
using SleipnirCommon.Models;

namespace Sleipnir.Samples.CSharp;

public static class SingleCallScenario
{
    public static async Task RunAsync(SleipnirRestJsonClient rest, TextWriter w)
    {
        // --- REST: Kunden anlegen (skalarer int kommt zurück) ----------------------
        var addReq = SleipnirCall.Init("Customer", "AddCustomer")
            .Param("name", "Alice")          // named parameter — server bindet nach Name
            .Param("email", "alice@x.com")
            .ToRequest();

        int newId = await rest.Call<int>(addReq);
        await w.WriteLineAsync($"  [REST]    AddCustomer  -> new id = {newId}");

        // --- REST: Kunden laden (Call<T> deserialisiert direkt nach Customer) ------
        var getReq = SleipnirCall.Init("Customer", "GetCustomerById")
            .Param("id", newId)
            .ToRequest();

        Customer? customer = await rest.Call<Customer>(getReq);
        await w.WriteLineAsync($"  [REST]    GetCustomerById({newId}) -> {customer?.Name} <{customer?.Email}>");

        // --- REST: alle Kunden (Liste) ---------------------------------------------
        var allReq = SleipnirCall.Init("Customer", "GetAllCustomers").ToRequest();
        var all = await rest.Call<List<Customer>>(allReq);
        await w.WriteLineAsync($"  [REST]    GetAllCustomers -> {all?.Count ?? 0} customer(s)");

        // --- WebSocket: derselbe Aufruf über den persistenten Kanal ----------------
        // IAsyncDisposable. Der C#-WS-Client nimmt die Basis-URL (https://…) + einen
        // separaten wsPath (Default "sleipnirws") und hebt intern auf wss://…/sleipnirws ab.
        await using var ws = new SleipnirWebSocketClient("https://localhost:5001");
        await ws.ConnectAsync();

        // With(params object?[]) = positionsbasiert (Namen param0, param1, …). Der
        // Server bindet positional, wenn der Name nicht passt — für Ein-Parameter-
        // Methoden ok.
        var wsReq = SleipnirCall.Init("Customer", "GetCustomerById")
            .With(newId)
            .ToRequest();

        Customer? customer2 = await ws.Call<Customer>(wsReq);
        await w.WriteLineAsync($"  [WebSocket] GetCustomerById({newId}) -> {customer2?.Name} <{customer2?.Email}>");

        // --- Raw-Form (ohne Fluent Builder) — gelegentlich nützlich ----------------
        // Params = nativer JSON-Wert: JsonArray aus { parameterName, data }-Einträgen,
        // data ist der JSON-Token selbst (Zahl 42 → 42, String "A" → "A").
        var raw = new SleipnirRequest
        {
            Controller = "Customer",
            Method = "GetAllCustomers",
            Id = "Customer.GetAllCustomers",
        };
        var rawList = await rest.Call<List<Customer>>(raw);
        await w.WriteLineAsync($"  [REST raw] GetAllCustomers -> {rawList?.Count ?? 0} customer(s)");
    }
}